File: tests/api_test.php
<?php

declare(strict_types=1);

use KpqC\KeyEncapsulationAlgorithm;
use KpqC\SignatureAlgorithm;
use function KpqC\aimer128f;
use function KpqC\aimer128s;
use function KpqC\aimer192f;
use function KpqC\aimer192s;
use function KpqC\aimer256f;
use function KpqC\aimer256s;
use function KpqC\haetae2;
use function KpqC\haetae3;
use function KpqC\haetae5;
use function KpqC\ntruplus768;
use function KpqC\ntruplus864;
use function KpqC\ntruplus1152;
use function KpqC\smaugt128;
use function KpqC\smaugt192;
use function KpqC\smaugt256;
use function KpqC\timer;

function check(bool $condition, string $message): void
{
    if (!$condition) {
        throw new RuntimeException($message);
    }
}

function changed(string $value): string
{
    return chr(ord($value[0]) ^ 0x80) . substr($value, 1);
}

/** @param class-string<Throwable> $expected */
function expectException(string $expected, callable $operation): void
{
    try {
        $operation();
    } catch (Throwable $error) {
        check(
            $error instanceof $expected,
            sprintf('expected %s, received %s', $expected, $error::class),
        );
        return;
    }
    throw new RuntimeException("expected {$expected}");
}

$signatures = [
    aimer128f(), aimer128s(), aimer192f(), aimer192s(),
    aimer256f(), aimer256s(), haetae2(), haetae3(), haetae5(),
];
$kems = [
    ntruplus768(), ntruplus864(), ntruplus1152(),
    smaugt128(), smaugt192(), smaugt256(), timer(),
];

foreach ($signatures as $algorithm) {
    check($algorithm instanceof SignatureAlgorithm, "{$algorithm->id}: type");
    $keys = $algorithm->generateKeyPair();
    $other = $algorithm->generateKeyPair();
    check(strlen($keys->publicKey) === $algorithm->sizes->publicKey, "{$algorithm->id}: public key size");
    check(strlen($keys->secretKey) === $algorithm->sizes->secretKey, "{$algorithm->id}: secret key size");
    check($keys->publicKey !== $other->publicKey, "{$algorithm->id}: fresh key pair");

    foreach ([["test\0\xff", 'domain'], ['', ''], ['boundary', str_repeat('x', 255)]] as [$message, $context]) {
        $signature = $algorithm->sign($message, $keys->secretKey, $context);
        check(strlen($signature) === $algorithm->sizes->signature, "{$algorithm->id}: signature size");
        check($algorithm->verify($message, $signature, $keys->publicKey, $context), "{$algorithm->id}: verification");
        check(!$algorithm->verify($message . '!', $signature, $keys->publicKey, $context), "{$algorithm->id}: changed message");
        check(!$algorithm->verify($message, changed($signature), $keys->publicKey, $context), "{$algorithm->id}: changed signature");
        check(!$algorithm->verify($message, $signature, $other->publicKey, $context), "{$algorithm->id}: wrong public key");
        check(!$algorithm->verify($message, $signature, $keys->publicKey, 'wrong'), "{$algorithm->id}: wrong context");
    }

    check(!$algorithm->verify('', '', $keys->publicKey), "{$algorithm->id}: wrong signature size");
    expectException(InvalidArgumentException::class, fn () => $algorithm->sign('', $keys->secretKey, str_repeat('x', 256)));
    expectException(InvalidArgumentException::class, fn () => $algorithm->sign('', substr($keys->secretKey, 1)));
    expectException(InvalidArgumentException::class, fn () => $algorithm->verify('', '', substr($keys->publicKey, 1)));
}

foreach ($kems as $algorithm) {
    check($algorithm instanceof KeyEncapsulationAlgorithm, "{$algorithm->id}: type");
    $keys = $algorithm->generateKeyPair();
    $first = $algorithm->encapsulate($keys->publicKey);
    $second = $algorithm->encapsulate($keys->publicKey);
    check(strlen($keys->publicKey) === $algorithm->sizes->publicKey, "{$algorithm->id}: public key size");
    check(strlen($keys->secretKey) === $algorithm->sizes->secretKey, "{$algorithm->id}: secret key size");
    check(strlen($first->ciphertext) === $algorithm->sizes->ciphertext, "{$algorithm->id}: ciphertext size");
    check(strlen($first->sharedSecret) === $algorithm->sizes->sharedSecret, "{$algorithm->id}: secret size");
    check($algorithm->decapsulate($first->ciphertext, $keys->secretKey) === $first->sharedSecret, "{$algorithm->id}: decapsulation");
    check($first->ciphertext !== $second->ciphertext, "{$algorithm->id}: fresh ciphertext");
    check($first->sharedSecret !== $second->sharedSecret, "{$algorithm->id}: fresh shared secret");

    if (str_starts_with($algorithm->id, 'NTRU+')) {
        expectException(Exception::class, fn () => $algorithm->decapsulate(changed($first->ciphertext), $keys->secretKey));
        expectException(Exception::class, fn () => $algorithm->encapsulate(str_repeat("\xff", $algorithm->sizes->publicKey)));
    } else {
        $rejected = $algorithm->decapsulate(changed($first->ciphertext), $keys->secretKey);
        check(strlen($rejected) === $algorithm->sizes->sharedSecret, "{$algorithm->id}: implicit rejection size");
        check($rejected !== $first->sharedSecret, "{$algorithm->id}: implicit rejection value");
    }

    expectException(InvalidArgumentException::class, fn () => $algorithm->encapsulate(substr($keys->publicKey, 1)));
    expectException(InvalidArgumentException::class, fn () => $algorithm->decapsulate(substr($first->ciphertext, 1), $keys->secretKey));
    expectException(InvalidArgumentException::class, fn () => $algorithm->decapsulate($first->ciphertext, substr($keys->secretKey, 1)));
}

$keyText = (string) ntruplus768()->generateKeyPair();
$secretText = (string) ntruplus768()->encapsulate(ntruplus768()->generateKeyPair()->publicKey);
check(str_contains($keyText, '[REDACTED]'), 'key string redacts the secret key');
check(str_contains($secretText, '[REDACTED]'), 'encapsulation string redacts the shared secret');

check(
    (new ReflectionClass(SignatureAlgorithm::class))->isInternal(),
    'signature API is native',
);
check(
    (new ReflectionFunction('KpqC\\aimer128f'))->isInternal(),
    'algorithm factories are native',
);
check(!function_exists('kpqc_native_metadata'), 'internal bridge API is absent');
check(!function_exists('kpqc_native_test_seed'), 'test entropy is absent from production');

check(count($signatures) === 9, 'signature algorithm count');
check(count($kems) === 7, 'KEM algorithm count');

echo "All PHP API checks passed across 16 parameter sets.\n";

File: native/kpqc/tests/api_test.cpp
#include "kpqc/kpqc.hpp"

#include <array>
#include <functional>
#include <iostream>
#include <stdexcept>
#include <string>

namespace {

void require(bool condition, const std::string& label) {
    if (!condition) {
        throw std::runtime_error(label + " failed");
    }
}

void expect_invalid_argument(
    const std::function<void()>& action,
    const std::string& label) {
    try {
        action();
    } catch (const std::invalid_argument&) {
        return;
    }
    throw std::runtime_error(label + " did not reject invalid input");
}

void expect_native_error(
    const std::function<void()>& action,
    const std::string& label) {
    try {
        action();
    } catch (const kpqc::Error&) {
        return;
    }
    throw std::runtime_error(label + " did not report a native failure");
}

struct KemCase {
    const kpqc::KeyEncapsulationAlgorithm* algorithm;
    bool implicit_rejection;
};

void check_signature(const kpqc::SignatureAlgorithm& algorithm) {
    const kpqc::KeyPair keys = algorithm.generate_key_pair();
    require(keys.public_key.size() == algorithm.sizes().public_key,
            std::string(algorithm.id()) + " public-key size");
    require(keys.secret_key.size() == algorithm.sizes().secret_key,
            std::string(algorithm.id()) + " secret-key size");

    const kpqc::Bytes message{'K', 'p', 'q', 'C', ' ', 'C', '+', '+'};
    const kpqc::Bytes context{'r', 'e', 'l', 'e', 'a', 's', 'e'};
    const kpqc::Bytes signature = algorithm.sign(message, keys.secret_key, context);
    require(signature.size() == algorithm.sizes().signature,
            std::string(algorithm.id()) + " signature size");
    require(algorithm.verify(message, signature, keys.public_key, context),
            std::string(algorithm.id()) + " verification");

    kpqc::Bytes altered_signature = signature;
    altered_signature.front() ^= 1U;
    require(!algorithm.verify(message, altered_signature, keys.public_key, context),
            std::string(algorithm.id()) + " altered-signature rejection");
    require(!algorithm.verify(message, kpqc::Bytes{0}, keys.public_key, context),
            std::string(algorithm.id()) + " short-signature rejection");
    require(!algorithm.verify(message, signature, keys.public_key, {'x'}),
            std::string(algorithm.id()) + " wrong-context rejection");
}

void check_kem(const KemCase& test) {
    const kpqc::KeyEncapsulationAlgorithm& algorithm = *test.algorithm;
    const kpqc::KeyPair keys = algorithm.generate_key_pair();
    require(keys.public_key.size() == algorithm.sizes().public_key,
            std::string(algorithm.id()) + " public-key size");
    require(keys.secret_key.size() == algorithm.sizes().secret_key,
            std::string(algorithm.id()) + " secret-key size");
    const kpqc::EncapsulatedSecret outbound =
        algorithm.encapsulate(keys.public_key);
    require(outbound.ciphertext.size() == algorithm.sizes().ciphertext,
            std::string(algorithm.id()) + " ciphertext size");
    require(outbound.shared_secret.size() == algorithm.sizes().shared_secret,
            std::string(algorithm.id()) + " shared-secret size");
    require(algorithm.decapsulate(outbound.ciphertext, keys.secret_key) ==
                outbound.shared_secret,
            std::string(algorithm.id()) + " decapsulation");

    kpqc::Bytes altered_ciphertext = outbound.ciphertext;
    altered_ciphertext.front() ^= 1U;
    if (test.implicit_rejection) {
        const kpqc::Bytes replacement =
            algorithm.decapsulate(altered_ciphertext, keys.secret_key);
        require(replacement != outbound.shared_secret,
                std::string(algorithm.id()) + " implicit rejection");
    } else {
        expect_native_error(
            [&algorithm, &altered_ciphertext, &keys] {
                static_cast<void>(algorithm.decapsulate(
                    altered_ciphertext, keys.secret_key));
            },
            std::string(algorithm.id()) + " altered-ciphertext rejection");
        expect_native_error(
            [&algorithm] {
                static_cast<void>(algorithm.encapsulate(
                    kpqc::Bytes(algorithm.sizes().public_key, 0xff)));
            },
            std::string(algorithm.id()) + " non-canonical public-key rejection");
    }

    expect_invalid_argument(
        [&algorithm, &keys] {
            static_cast<void>(algorithm.decapsulate(
                kpqc::Bytes(algorithm.sizes().ciphertext - 1),
                keys.secret_key));
        },
        std::string(algorithm.id()) + " short ciphertext");
    expect_invalid_argument(
        [&algorithm, &outbound] {
            static_cast<void>(algorithm.decapsulate(
                outbound.ciphertext,
                kpqc::Bytes(algorithm.sizes().secret_key - 1)));
        },
        std::string(algorithm.id()) + " short secret key");
}

}  // namespace

int main() {
    try {
        const std::array<const kpqc::SignatureAlgorithm*, 9> signatures{{
            &kpqc::aimer::aimer128f(), &kpqc::aimer::aimer128s(),
            &kpqc::aimer::aimer192f(), &kpqc::aimer::aimer192s(),
            &kpqc::aimer::aimer256f(), &kpqc::aimer::aimer256s(),
            &kpqc::haetae::haetae2(), &kpqc::haetae::haetae3(),
            &kpqc::haetae::haetae5(),
        }};
        const std::array<KemCase, 7> kems{{
            {&kpqc::ntruplus::ntruplus768(), false},
            {&kpqc::ntruplus::ntruplus864(), false},
            {&kpqc::ntruplus::ntruplus1152(), false},
            {&kpqc::smaugt::smaugt128(), true},
            {&kpqc::smaugt::smaugt192(), true},
            {&kpqc::smaugt::smaugt256(), true},
            {&kpqc::smaugt::timer(), true},
        }};

        for (const kpqc::SignatureAlgorithm* algorithm : signatures) {
            check_signature(*algorithm);
        }
        for (const KemCase& test : kems) {
            check_kem(test);
        }

        const kpqc::SignatureAlgorithm& signature =
            kpqc::aimer::aimer128f();
        const kpqc::KeyPair signature_keys = signature.generate_key_pair();
        expect_invalid_argument(
            [&signature] { signature.sign({}, {0}); }, "short secret key");
        expect_invalid_argument(
            [&signature, &signature_keys] {
                signature.sign({}, signature_keys.secret_key, kpqc::Bytes(256));
            },
            "long signature context");

        const kpqc::KeyEncapsulationAlgorithm& kem =
            kpqc::ntruplus::ntruplus768();
        expect_invalid_argument(
            [&kem] { kem.encapsulate({0}); }, "short public key");

        std::cout << "C++ API behavior checks passed across 16 parameter sets.\n";
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "API failure: " << error.what() << '\n';
        return 1;
    }
}
